src/mgmt_conn.c

```diff
@@ -717,6 +717,174 @@ internal_snapshot(char *snap)
 	return (B_FALSE);
 }
 
+/*
+ * This fn returns map of snapshots by walking through datasets in json
+ * format. Returns 0 on success.
+ * This will return EAGAIN if the replica is NOT connected to target.
+ */
+static struct json_object *
+uzfs_zvol_fetch_snapshot_jmap(zvol_info_t *zinfo, int *err)
+{
+	char *snapname;
+	boolean_t case_conflict;
+	uint64_t id, pos = 0;
+	int error = 0;
+	zvol_state_t *zv = (zvol_state_t *)zinfo->main_zv;
+	objset_t *os = zv->zv_objset;
+	dsl_dataset_t *ds;
+	dsl_pool_t *dp;
+	struct json_object *jmap = NULL;
+
+	snapname = kmem_zalloc(ZFS_MAX_DATASET_NAME_LEN, KM_SLEEP);
+
+	if ((os == NULL) || (os->os_dsl_dataset == NULL) ||
+	    (os->os_dsl_dataset->ds_dir == NULL) ||
+	    (os->os_dsl_dataset->ds_dir->dd_pool == NULL)) {
+		error = EAGAIN;
+		goto out;
+	}
+
+	jmap = json_object_new_object();
+	dp = os->os_dsl_dataset->ds_dir->dd_pool;
+
+	while (error == 0) {
+		dsl_pool_config_enter(dmu_objset_pool(os), FTAG);
+		error = dmu_snapshot_list_next(os,
+		    ZFS_MAX_DATASET_NAME_LEN, snapname, &id, &pos,
+		    &case_conflict);
+		if (error) {
+			dsl_pool_config_exit(dmu_objset_pool(os), FTAG);
+			break;
+		}
+
+		if (internal_snapshot(snapname)) {
+			dsl_pool_config_exit(dmu_objset_pool(os), FTAG);
+			continue;
+		}
+
+		dsl_pool_config_exit(dmu_objset_pool(os), FTAG);
+
+		json_object_object_add(jmap, snapname, NULL);
+	}
+
+	if ((error == ENOENT) || (error == 0)) {
+		error = 0;
+	} else {
+		json_object_put(jmap);
+		jmap = NULL;
+	}
+out:
+	kmem_free(snapname, ZFS_MAX_DATASET_NAME_LEN);
+	*err = error;
+
+	return (jmap);
+}
+
+/*
+ * This fn returns reference to zinfo's map of snap json_object.
+ * If its NULL, it creates one and returns it
+ * Its the responsibility of caller to release the reference.
+ * Also, this fn need to be called with snap_map_mutex.
+ */
+int
+uzfs_zvol_get_snap_jmap_ref(zvol_info_t *zinfo, struct json_object **jmapp)
+{
+	struct json_object *jmap;
+	int error = 0;
+
+	ASSERT(MUTEX_HELD(&zinfo->snap_map_mutex));
+	*jampp = NULL;
+	if (zinfo->snap_map == NULL) {
+
+		jmap = uzfs_zvol_fetch_snapshot_jmap(zinfo, &error);
+
+		if (error)
+			return (error);
+
+		zinfo->snap_map = jmap;
+	}
+	jmap = json_object_get(zinfo->snap_map);
+
+	*jmapp = jmap;
+	return (0);
+}
+
+/*
+ * This fn adds/deletes snapname to zinfo's snap_map in json format.
+ * If errors, sets snap_map to NULL as update failed
+ * This fn is currently called only if snapshot is created / deleted through
+ * istgt, but, not through CLI
+ */
+int
+uzfs_zvol_update_snapshot_jmap(zvol_info_t *zinfo, char *snapname,
+    boolean_t add)
+{
+	int ret = 0;
+	struct json_object *jmap;
+
+	pthread_mutex_lock(&zinfo->snap_map_mutex);
+	ret = uzfs_zvol_get_snap_jmap_ref(zinfo, &jmap);
+	if (ret == 0) {
+		if (add)
+			json_object_object_add(jmap, snapname, NULL);
+		else
+			json_object_object_del(jmap, snapname);
+		json_object_put(jmap);
+	} else {
+		if (zinfo->snap_map != NULL)
```

Comment:
Should we add assertion here for `zinfo->snap_map` as it should be `NULL`?

Reply:
sometimes it might not be nil as well